complex/ft_main.c

```diff
@@ -253,11 +254,94 @@ static int ft_fw_result_index(int fi_errno)
 	}
 }
 
-static int ft_fw_server(void)
+static int ft_fw_process_list(struct fi_info *hints, struct fi_info *info)
+{
+	int ret, subindex, result;
+	size_t len;
+
+	for (subindex = 1, fabric_info = info; fabric_info;
+	     fabric_info = fabric_info->next, subindex++) {
+
+		//check needed to skip utility providers, unless requested
+		if (!ft_util_name(hints->fabric_attr->prov_name, &len) &&
+		    strcmp(hints->fabric_attr->prov_name,
+		    fabric_info->fabric_attr->prov_name))
+			continue;
+
+		ret = ft_check_info(hints, fabric_info);
+		if (ret)
+			return ret;
+
+		ft_fw_update_info(&test_info, fabric_info, subindex);
+		printf("Starting test %d-%d: ", test_info.test_index,
+			subindex);
+		ft_show_test_info();
+
+		result = ft_run_test();
+		results[ft_fw_result_index(-result)]++;
+
+		ret = ft_sock_send(sock, &result, sizeof result);
+		if (result) {
+			FT_PRINTERR("ft_run_test", result);
+		} else if (ret) {
+			FT_PRINTERR("ft_sock_send", ret);
+			return ret;
+		}
+	}
+	test_info.prov_name[0] = '\0';
+	ret = ft_sock_send(sock, &test_info, sizeof test_info);
+	if (ret) {
+		FT_PRINTERR("ft_sock_send", ret);
+		return ret;
+	}
+
+	if (subindex == 1)
+		return -FI_ENODATA;
+
+	return 0;
+}
+
+static int ft_server_child()
 {
 	struct fi_info *hints, *info;
 	int ret;
 
+	hints = fi_allocinfo();
+	if (!hints)
+		return -FI_ENOMEM;
+
+	ft_fw_convert_info(hints, &test_info);
+	printf("Starting test %d:\n", test_info.test_index);
+
+	ret = fi_getinfo(FT_FIVERSION, ft_strptr(test_info.node),
+			 ft_strptr(test_info.service), FI_SOURCE,
+			 hints, &info);
+	if (ret && ret != -FI_ENODATA) {
+		FT_PRINTERR("fi_getinfo", ret);
```

Comment:
The else actually needs to contain the ENODATA because the process_list function is what signals the client that the test is over, so it needs to enter that function in order to terminate that test. Or I can create a function that would send the termination signal and call that in both process_list and also after getinfo returns ENODATA. Preference?